### ISSUE TITLE
Convolutional Sequence to Sequence

### ISSUE THREAD

**Issue by kuldeep7688** [bentrevett/pytorch-seq2seq#38]
In both encoder and decoder we there multiple layers to convert emb_dim to hid_dim and vice versa. There are similar linear layers for attention too. 

Can you please explain a little bit about why it is being used? The intuition behind it basically.

**Comment by bentrevett** [comment]
I'll quote the relevant paragraph from the paper here:

"We also add linear mappings to project between the embedding size `f` and the convolution outputs that are of size `2d`. We apply such a transform to `w` when feeding embeddings to the encoder network, to the encoder output `z^u_j`, to the final layer of the decoder just before the softmax `h^L` and to all decoder layers, `h^l` before computing attention scores`

`f` is the embedding dimension, `d` is the hidden dimension and `w` are the embedded tokens. The rest are explained in the quote.

I don't know their intuition behind it as they don't state it in the paper, but I'll try and state why I think they did it. The `emb2hid` layers will learn how to project the embeddings into a "better" space to be operated on by the convolutional blocks. This allows the embedding layer to focus only on getting sensible token embeddings and not have to worry about how they should be before feeding them into the conv. blocks. This way, when you do the residual sum to get the `combined` output you have one part from the embeddings (which is an embedding of that token and that token only) summed with one part from the conv. blocks (which is made up of that token AND other tokens in the sequence). 

Apparently separating these things out seemed to work for them - I'm sure they will have tried without these linear transformation (`emb2hid` and `hid2emb`) layers and putting the embeddings straight into the conv. blocks and I guess it didn't work as well? 

One test would be to remove those layers, change the appropriate code and see how well it does. If you do that, let me know how it goes!

**Comment by kuldeep7688** [comment]
Yeah, I think your intuition is correct. 
Definitely I will try without these linear transformations (emb2hid and hid2emb) and will let you know about results.

Thanks again for such a great tutorial series on seq2seq, it helped me a lot.
